File: DPBO-praktikum/4/python/Class.py
class Human:
    def __init__(self, nik, nama, jenis_kelamin):
        self._nik = nik
        self._nama = nama
        self._jenis_kelamin = jenis_kelamin

    def get_nik(self):
        return self._nik

    def set_nik(self, nik):
        self._nik = nik

    def get_nama(self):
        return self._nama

    def set_nama(self, nama):
        self._nama = nama

    def get_jenis_kelamin(self):
        return self._jenis_kelamin

    def set_jenis_kelamin(self, jenis_kelamin):
        self._jenis_kelamin = jenis_kelamin


class SivitasAkademik(Human):
    def __init__(self, nik, nama, jenis_kelamin, asal_universitas, email_edu):
        super().__init__(nik, nama, jenis_kelamin)
        self._asal_universitas = asal_universitas
        self._email_edu = email_edu

    def set_asal_universitas(self, asal_universitas):
        self._asal_universitas = asal_universitas

    def get_asal_universitas(self):
        return self._asal_universitas

    def set_email_edu(self, email_edu):
        self._email_edu = email_edu

    def get_email_edu(self):
        return self._email_edu

class Mahasiswa(SivitasAkademik):
    def __init__(self, nim, nama, jenis_kelamin, fakultas, asal_universitas, email_edu):
        super().__init__(None, nama, jenis_kelamin, asal_universitas, email_edu)
        self._nim = nim
        self._fakultas = fakultas
        self._prodi = None

    def get_nim(self):
        return self._nim

    def set_nim(self, nim):
        self._nim = nim

    def get_fakultas(self):
        return self._fakultas

    def set_fakultas(self, fakultas):
        self._fakultas = fakultas

    def get_prodi(self):
        return self._prodi

    def set_prodi(self, prodi):
        self._prodi = prodi


class Dosen(SivitasAkademik):
    def __init__(self, nik, nip, nama, jenis_kelamin, fakultas, pend_terakhir, keahlian, asal_universitas, email_edu):
        super().__init__(nik, nama, jenis_kelamin, asal_universitas, email_edu)
        self._nip = nip
        self._fakultas = fakultas
        self._pend_terakhir = pend_terakhir
        self._keahlian = keahlian

    def get_nip(self):
        return self._nip

    def set_nip(self, nip):
        self._nip = nip

    def get_fakultas(self):
        return self._fakultas

    def set_fakultas(self, fakultas):
        self._fakultas = fakultas

    def get_pend_terakhir(self):
        return self._pend_terakhir

    def set_pend_terakhir(self, pend_terakhir):
        self._pend_terakhir = pend_terakhir

    def get_keahlian(self):
        return self._keahlian

    def set_keahlian(self, keahlian):
        self._keahlian = keahlian


class Course:
    def __init__(self, nama_matakuliah, dosen):
        self._nama_matakuliah = nama_matakuliah
        self._dosen = dosen
        self._mahasiswa = []

    def get_nama_matakuliah(self):
        return self._nama_matakuliah
    
    def set_nama_matakuliah(self, nama_matakuliah):
        self._nama_matakuliah = nama_matakuliah

    def get_dosen(self):
        return self._dosen

    def set_dosen(self, dosen):
        self._dosen = dosen

    def get_mahasiswa(self):
        return self._mahasiswa

    def add_mahasiswa(self, mahasiswa):
        self._mahasiswa.append(mahasiswa)

    def remove_mahasiswa(self, mahasiswa):
        if mahasiswa in self._mahasiswa:
            self._mahasiswa.remove(mahasiswa)

class ProgramStudi:
    def __init__(self, nama_prodi, kode):
        self._nama_prodi = nama_prodi
        self._kode = kode
        self._courses = []

    def get_nama_prodi(self):
        return self._nama_prodi

    def set_nama_prodi(self, nama_prodi):
        self._nama_prodi = nama_prodi

    def get_kode(self):
        return self._kode

    def set_kode(self, kode):
        self._kode = kode

    def get_courses(self):
        return self._courses

    def add_course(self, course):
        self._courses.append(course)

    def remove_course(self, course):
        if course in self._courses:
            self._courses.remove(course)

File: DPBO-praktikum/4/cpp/Class.cpp
// #include <iostream>
// #include<vector>
// #include<string>
// using namespace std;

// class Human {
//     private:
//     string nik;
//     string nama;
//     string jenis_kelamin;

//     public:
//     Human(){

//     }

//     Human(string nik, string nama, string jenis_kelamin) {
//     this->nik = nik;
//     this->nama = nama;
//     this->jenis_kelamin = jenis_kelamin;
//     }
//     string get_nik() {
//         return nik;
//     }

//     void set_nik(string nik) {
//         this->nik = nik;
//     }

//     string get_nama() {
//         return nama;
//     }

//     void set_nama(string nama) {
//         this->nama = nama;
//     }

//     string get_jenis_kelamin() {
//         return jenis_kelamin;
//     }

//     void set_jenis_kelamin(string jenis_kelamin) {
//         this->jenis_kelamin = jenis_kelamin;
//     }

//     ~Human(){

//     }
// };

// class SivitasAkademik : public Human {
//     private:
//     string asal_universitas;
//     string email_edu;

//     public:
//     SivitasAkademik(){

//     }

//     SivitasAkademik(string nik, string nama, string jenis_kelamin, string asal_universitas, string email_edu)
//     : Human(nik, nama, jenis_kelamin) {
//     this->asal_universitas = asal_universitas;
//     this->email_edu = email_edu;
//     }
//     void set_asal_universitas(string asal_universitas) {
//         this->asal_universitas = asal_universitas;
//     }

//     string get_asal_universitas() {
//         return asal_universitas;
//     }

//     void set_email_edu(string email_edu) {
//         this->email_edu = email_edu;
//     }

//     string get_email_edu() {
//         return email_edu;
//     }
//     ~SivitasAkademik(){

//     }
// };

// class Mahasiswa : public SivitasAkademik {
//     private:
//     string nim;
//     string fakultas;
//     string prodi;

//     public:
//     Mahasiswa(string nik,string nim, string nama, string jenis_kelamin, string fakultas, string asal_universitas, string email_edu)
//     : SivitasAkademik(nik, nama, jenis_kelamin, asal_universitas, email_edu) {
//     this->nim = nim;
//     this->fakultas = fakultas;
//     this->prodi = "";
//     }
//     string get_nim() {
//         return nim;
//     }

//     void set_nim(string nim) {
//         this->nim = nim;
//     }

//     string get_fakultas() {
//         return fakultas;
//     }

//     void set_fakultas(string fakultas) {
//         this->fakultas = fakultas;
//     }

//     string get_prodi() {
//         return prodi;
//     }

//     void set_prodi(string prodi) {
//         this->prodi = prodi;
//     }
//     ~Mahasiswa(){

//     }
// };

// class Dosen : public SivitasAkademik {
//     private:
//         string nip;
//         string fakultas;
//         string pend_terakhir;
//         string keahlian;
//     public:
//         Dosen(){
            
//         }

//         Dosen(string nik, string nip, string nama, string jenis_kelamin, string fakultas, string pend_terakhir, string keahlian, string asal_universitas, string email_edu) : SivitasAkademik("", nama, jenis_kelamin, asal_universitas, email_edu) {
//             this->nip = nip;
//             this->fakultas = fakultas;
//             this->pend_terakhir=pend_terakhir;
//             this->keahlian=keahlian;
//         };
        
//         string get_nip(){
//             return nip;
//         };
//         void set_nip(string nip){
//             this->nip = nip;
//         };

//         string get_fakultas(){
//             return fakultas;
//         };
//         void set_fakultas(string fakultas){
//             this->fakultas = fakultas;
//         };

//         string get_pend_terakhir(){
//             return pend_terakhir;
//         };
//         void set_pend_terakhir(string pend_terakhir){
//             this->pend_terakhir = pend_terakhir;
//         };
//         string get_keahlian(){
//             return keahlian;
//         };
//         void set_keahlian(string keahlian){
//             this->keahlian = keahlian;
//         };
// };

// class Course {
//     private:
//         string nama_matakuliah;
//         Dosen dosen;
//         vector<Mahasiswa> mahasiswa;

//     public:
//         Course(string nama_matakuliah, Dosen dosen) {
//             this->nama_matakuliah = nama_matakuliah;
//             this->dosen = dosen;
//         }

//         ~Course() {
//             mahasiswa.clear();
//         }

//         string get_nama_matakuliah() {
//             return this->nama_matakuliah;
//         }

//         void set_nama_matakuliah(string nama_matakuliah) {
//             this->nama_matakuliah = nama_matakuliah;
//         }

//         Dosen get_dosen() {
//             return this->dosen;
//         }

//         void set_dosen(Dosen dosen) {
//             this->dosen = dosen;
//         }

//         vector<Mahasiswa> get_mahasiswa() {
//             return this->mahasiswa;
//         }

//         void add_mahasiswa(Mahasiswa mahasiswa) {
//             this->mahasiswa.push_back(mahasiswa);
//         }
// };


// class ProgramStudi {
// private:
//     string nama_prodi;
//     string kode;
//     vector<Course> courses; // tidak menggunakan pointer
// public:
//     ProgramStudi(string nama_prodi, string kode) : nama_prodi(nama_prodi), kode(kode) {}

//     string get_nama_prodi() {
//         return nama_prodi;
//     }

//     void set_nama_prodi(string nama_prodi) {
//         this->nama_prodi = nama_prodi;
//     }

//     string get_kode() {
//         return kode;
//     }

//     void set_kode(string kode) {
//         this->kode = kode;
//     }

//     vector<Course> get_courses() {
//         return courses;
//     }

//     void add_course(Course course) { // bukan pointer
//         courses.push_back(course);
//     }

//     ~ProgramStudi() { // tidak perlu dealokasi karena tidak menggunakan pointer
//     }
// };
